Темизация

Winterstein9 opened the pull request:
### Какую задачу решаем
sta-29-themization
Добавление / изменение темы. Изменение темы доступно в профиле. Тема автоматически устанавливается в соответствии с темой устройства пользователя.
### Скриншоты/видяшка (если есть)

### TBD (если есть)

Winterstein9 committed:
1

Bluorenge commented on packages/client/src/components/ThemeButton/index.tsx:
```diff
@@ -0,0 +1,23 @@
+  const { theme, setTheme } = useTheme();
+
+  const themeSwitcher = () => {
+    console.log('hello');
```
консоль забыл убрать)

Bluorenge commented on packages/client/src/components/ThemeButton/ThemeButton.scss:
```diff
@@ -0,0 +1,14 @@
+#theme__button {
```
почему по айди обращаешься?
и ещё у нас в проекте именование по бэму такое - сущностьКчемуОтносится, в данном случае btnTheme.

Winterstein9 replied:
> почему по айди обращаешься? и ещё у нас в проекте именование по бэму такое - сущностьКчемуОтносится, в данном случае btnTheme.

id если стиль применяется к одному элементу, class если к более чем одному ...

Bluorenge replied:
Во-первых, у нас на проекте используются классы.
Во-вторых, проблема с уникальностью селектора решается css-модулями (но мы их почему-то сразу не использовали).

Winterstein9 committed:
1.1

Winterstein9 committed:
1.2

Bluorenge commented on packages/client/src/components/ThemeButton/index.tsx:
```diff
@@ -0,0 +1,22 @@
+import './ThemeButton.scss';
```
смотри как у нас юзаются компоненты - в index находится только экспорт

Winterstein9 replied:
> смотри как у нас юзаются компоненты - в index находится только экспорт

Для чего так делать?

Bluorenge replied:
для единообразия

Winterstein9 committed:
1.3

Bluorenge commented on packages/client/src/hooks/useTheme.ts:
```diff
@@ -0,0 +1,15 @@
+    localStorage.setItem('app-theme', theme);
+  }, [theme]);
+
+  return { theme, setTheme };
```
опять же ради единообразия (как в useLocale) давай как массив возвращать значения

Bluorenge replied:
да и дефолтные хуки реакта (юсСтейт) тоже как массив же

Bluorenge commented on packages/client/src/components/ThemeButton/ThemeButton.scss:
```diff
@@ -0,0 +1,14 @@
+.btnTheme {
+  background-color: white;
+  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto,
```
зачем здесь семейство шрифта?

Winterstein9 replied:
> зачем здесь семейство шрифта?

 Иначе кнопка будет Times New Roman По другому нужно подключить? Как?

Bluorenge replied:
вообще надо было использовать компонент button из antd

Bluorenge replied:
а вдруг семейство шрифтов изменится? можно забыть поменять здесь шрифт

Bluorenge commented on packages/client/src/components/Profile/ProfileForm/ProfileForm.tsx:
```diff
@@ -225,20 +227,25 @@ export const ProfileForm = () => {
           {fm(messages.buttonChangePassword)}
         </Button>
+
+        <ThemeButton />
```
Это надо было сразу обсудить, но может в хедер добавим смену темы? Рядом со сменой языка.

К тому же то, что она здесь путает - кнопка находится рядом со ссылками на другие страницы и ожидание, что это тоже ссылка на какую-то страницу смены темы.

Bluorenge replied:
Что скажешь? Сейчас эта кнопка как ссылка неочевидна для пользователя

Winterstein9 replied:
> Что скажешь? Сейчас эта кнопка как ссылка неочевидна для пользователя

переделываю... фактически готово, но мешают ошибки отображения сайта ... наверное нужно браузер почистить

Bluorenge commented on packages/client/src/components/ThemeButton/index.tsx:
```diff
@@ -0,0 +1,22 @@
+    <input
+      className="btnTheme"
+      type="button"
+      value="Изменить тему"
```
Сюда бы переводы добавить сразу. Как это сделать можно посмотреть в других компонентах.

Bluorenge replied:
переводы решил не делать?

Winterstein9 committed:
1.4

Winterstein9 committed:
1.5

Winterstein9 committed:
sta-29-themization перемещение переключателя темы

Кнопка переключения темы перемещена из профиля пользователя на панель навигации. Изменен внешний вид кнопки.

Bluorenge commented on .hintrc:
```diff
@@ -0,0 +1,13 @@
+{
```
а этот файл зачем? я просто не в теме)

Winterstein9 replied:
> а этот файл зачем? я просто не в теме)

иначе ошибка отсутвия label хотя label есть и связь с input не меняет ситуации.

Bluorenge replied:
![image](https://user-images.githubusercontent.com/16067434/232495000-9af39ab9-369b-4d13-ac8a-3917bb1120df.png)
хм. у меня всё норм

Bluorenge commented on packages/client/src/components/ThemeButton/ThemeButton.tsx:
```diff
@@ -0,0 +1,28 @@
+    }
+  };
+  return (
+    <label className="switch">
```
окей, можно и так, но смотри, есть же свитчер в анте - https://ant.design/components/switch

Bluorenge commented on packages/client/src/components/ThemeButton/ThemeButton.tsx:
```diff
@@ -0,0 +1,28 @@
+    <label className="switch">
+      <input
+        type="checkbox"
+        value="light"
```
сюда надо передать theme, чтобы react ошибку не кидал в консоль https://react.dev/reference/react-dom/components/input#controlling-an-input-with-a-state-variable

Bluorenge commented on packages/client/src/components/ThemeButton/ThemeButton.tsx:
```diff
@@ -0,0 +1,28 @@
+
+  const checked = theme === 'dark' ? true : undefined;
+
+  const themeSwitcher = (e: any) => {
```
Лучше типизировать, чтобы ворнинга линтера не было. Передаём HTMLInputElement. Но тогда ниже нужно передавать e.currentTarget. Ну и аргумент тогда должен быть назван не e, а как-нибудь inputEl, потому что мы же не к событию обращаемся, а вызываем метод на элементе.

Bluorenge commented on packages/client/src/hooks/useTheme.ts:
```diff
@@ -0,0 +1,15 @@
+
+export const useTheme = () => {
+  const isDark = window?.matchMedia('(prefers-color-scheme:dark)');
+  const defaultTheme = isDark.matches ? 'dark' : 'light';
```
почему бы не типизировать? сделай по аналогии с useLocale. Там типизирован язык - есть импорт файла app.d.ts

Bluorenge commented on packages/client/src/hooks/useTheme.ts:
```diff
@@ -0,0 +1,15 @@
+  const [theme, setTheme] = useState(
+    localStorage.getItem('app-theme') || defaultTheme
+  );
+  useLayoutEffect(() => {
```
почему здесь useLayoutEffect? react не рекодмендует его использовать https://react.dev/reference/react/useLayoutEffect

Bluorenge commented on packages/client/src/hooks/useTheme.ts:
```diff
@@ -0,0 +1,15 @@
+  const isDark = window?.matchMedia('(prefers-color-scheme:dark)');
+  const defaultTheme = isDark.matches ? 'dark' : 'light';
+  const [theme, setTheme] = useState(
+    localStorage.getItem('app-theme') || defaultTheme
```
по аналогии с useLocale выдели app-theme в константу

Bluorenge replied:
и камелКейсом её сделай, чтобы было единообразно

Winterstein9 committed:
sta-29-themization

Внесение изменений в соответствии с предложениями Antony от 17 апреля 2023.

practicum-maxim-zyryanov commented on packages/client/src/hooks/useTheme.ts:
```diff
@@ -0,0 +1,18 @@
+import { useEffect, useState } from 'react';
+
+export const useTheme = () => {
+  const isDark: MediaQueryList = window?.matchMedia(
+    '(prefers-color-scheme:dark)'
+  );
```
Хорошая идея использовать `'(prefers-color-scheme:dark)'` в качестве дефолтного значения

practicum-maxim-zyryanov commented on packages/client/src/components/ThemeButton/ThemeButton.tsx:
```diff
@@ -0,0 +1,28 @@
+export const ThemeButton = () => {
+  const [theme, setTheme] = useTheme();
+
+  const checked = theme === 'dark' ? true : undefined;
+
+  const themeSwitcher = (inputEl: HTMLElement) => {
+    inputEl.blur();
+    if (theme === 'light') {
+      setTheme('dark');
+    } else {
+      setTheme('light');
+    }
+  };
```
Даже в этом небольшом фрагменте видим множество `dark` и `light`. Сделать опечатку вроде `daek` или `ligth` достаточно просто, а вот отыскать проблему будет сложно. Поэтому не помешало бы вынести эти строки в константы например.

practicum-maxim-zyryanov reviewed (approved):
Approve, но небольшое замечание оставил, да и хотелось бы видеть в описании скрины, в остальном спасибо в том числе @Bluorenge за проведённое ревью, реализация стала заметно лучше.

Winterstein9 committed:
sta-29-themization

Введены константы для тем.
